bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -2525,185 +2570,206 @@ void BCStateTran::processData(bool lastInBatch) {
+          ioContexts_.push_back(std::move(ctx));
+          histograms_.dst_num_pending_blocks_to_commit->record(ioContexts_.size());
+          finalizePutblockAsync(PutBlockWaitPolicy::NO_WAIT);
+          as_->getPrevDigestFromBlock(
+              buffer_.get(), actualBlockSize, reinterpret_cast<StateTransferDigest *>(&digestOfNextRequiredBlock_));
+          ConcordAssertGT(nextRequiredBlock_, 0);
+          --nextRequiredBlock_;
+          LOG_TRACE(logger_, KVLOG(nextRequiredBlock_));
+          if (lastInBatch || posponedSendFetchBlocksMsg_ || newSourceReplica) {
+            trySendFetchBlocksMsg(firstRequiredBlock,
+                                  nextRequiredBlock_,
+                                  0,
+                                  KVLOG(lastInBatch, posponedSendFetchBlocksMsg_, newSourceReplica));
+            break;
+          }
+        } else {
+          //////////////////////////////////////////////////////////////////////////
+          // This is the last block we need in cycle or on a fetch range
+          //////////////////////////////////////////////////////////////////////////
+          LOG_INFO(logger_, ss.str());
+          commitToChainDT_.start();
+          blocks_collected_.pause();
+          bytes_collected_.pause();
+          finalizePutblockAsync(PutBlockWaitPolicy::WAIT_ALL_JOBS, true);
+
+          ConcordAssertEQ(nextBlockIdToCommit_, nextRequiredBlock_);
+          ConcordAssert(ioContexts_.empty());
+          ConcordAssert(as_->putBlock(nextRequiredBlock_, buffer_.get(), actualBlockSize, true));
+
+          commitToChainDT_.pause();
+          digestOfNextRequiredBlock_.makeZero();
+          clearAllPendingItemsData();
+          DataStoreTransaction::Guard g(psd_->beginTransaction());
+          {
+            if (!lastFetchedBlockIdInCycle) {
+              const auto &lastSentMsg = std::get<FetchBlocksMsg>(lastSentMsg_);
+              ConcordAssertLE(lastSentMsg.maxBlockId + 1, psd_->getLastRequiredBlock());
+              g.txn()->setFirstRequiredBlock(lastSentMsg.maxBlockId + 1);
+              nextRequiredBlock_ = computeNextRequiredBlock();
+              nextBlockIdToCommit_ = nextRequiredBlock_;
+              trySendFetchBlocksMsg(psd_->getFirstRequiredBlock(),
+                                    nextRequiredBlock_,
+                                    0,
+                                    KVLOG(lastInBatch, posponedSendFetchBlocksMsg_, newSourceReplica));
+              break;
+            }
+            // case: blockPos == BlockPosition::LAST_FETCH_IN_CYCLE
+            nextRequiredBlock_ = 0;
+            nextBlockIdToCommit_ = 0;
+            g.txn()->setFirstRequiredBlock(0);
+            g.txn()->setLastRequiredBlock(0);
+          }
+          ConcordAssertEQ(getFetchingState(), FetchingState::GettingMissingResPages);
+
+          // Log histograms for destination when GettingMissingBlocks is done
```

Comment:
Should this be a separate function? maybe LogHistogram()?

Reply:
The code here is invalid anyway now. I could also comment it...all the statistics calculation are invalid.
We should do this as part of refactoring.